src/main/java/org/breedinginsight/brapps/importer/services/processors/experiment/create/workflow/CreateNewExperimentWorkflow.java

```diff
@@ -0,0 +1,307 @@
+        ProcessedData processedData = populateNewPendingImportObjectsStep.process(processContext, phenotypeData);
+        ValidationErrors validationErrors = validatePendingImportObjectsStep.process(context, processContext.getPendingData(), phenotypeData, processedData);
+
+        // short circuit if there were validation errors
+        if (validationErrors.hasErrors()) {
+            throw new ValidatorException(validationErrors);
+        }
+
+        // TODO: move to experiment import service
+        ImportPreviewResponse response = buildImportPreviewResponse(importRows, processContext.getPendingData(), processedData, upload);
+
+        statusService.updateMappedData(upload, response, "Finished mapping data to brapi objects");
+
+        // preview data
+        if (!commit) {
+            statusService.updateOk(upload);
+            return response;
+        }
+
+        // commit data
+        long totalObjects = getNewObjectCount(response);
+        statusService.startUpload(upload, totalObjects, "Starting upload to brapi service");
+        statusService.updateMessage(upload, "Creating new experiment objects in brapi service");
+
+        commitPendingImportObjectsStep.process(processContext, processedData);
+
+        statusService.finishUpload(upload, totalObjects, "Completed upload to brapi service");
+        return response;
+    }
+
+    /**
+     * Process the import service context and returns an Optional ImportWorkflowResult.
+     *
+     * @param context The import service context to be processed. If null, then it skips processing but returns the result with no-preview.
+     * @return An Optional ImportWorkflowResult which contains the workflow and import preview response (if available).
+     *         If the context is null, it returns the result with no-preview.
+     * @throws Exception If any error occurs during the processing.
+     */
+    public Optional<ImportWorkflowResult> process(ImportServiceContext context) throws Exception {
+        // Workflow processing the context
+        ImportWorkflow workflow = ImportWorkflow.builder()
+                .id(getWorkflow().getId())
+                .name(getWorkflow().getName())
+                .build();
+
+        // No-preview result
+        Optional<ImportWorkflowResult> result;
+
+        result = Optional.of(ImportWorkflowResult.builder()
+                .workflow(workflow)
+                .importPreviewResponse(Optional.empty())
+                .build());
+
+        // Skip this workflow unless creating a new experiment
+        if (context != null && !this.workflow.isEqual(context.getWorkflow())) {
+            return Optional.empty();
+        }
+
+        // Skip processing if no context, but return no-preview result for this workflow
+        if (context == null) {
+            return result;
+        }
+
+        // TODO: unify usage of single import context type throughout
+        ImportContext importContext = ImportContext.from(context);
```

Comment:
Agree. I suggest for `ImportServiceContext` renaming `workflow` to `workflowId`, `brapiImports` to `importRows`, and adding the `Map<Integer, PendingImport> mappedBrAPIImport` field. Then eliminating the `ImportContext` class.

Reply:
I created a card [BI-2230](https://breedinginsight.atlassian.net/browse/BI-2230?atlOrigin=eyJpIjoiOGVkNTFhMzcyMThkNGRjYTliYWNkNWI1N2FhOGRjNzciLCJwIjoiaiJ9) to do cleanup and reduce code duplication between workflows to limit the scope of changes in this PR.